### PR TITLE
core: only refuse Type=dbus service enqueue if dbus has pending stop job

### LINKED ISSUE #27588: core: fix use of manager_dbus_is_running()
This is just a reminder.
53964fd26b4a01191609ffc064aa8ccccd28e377 (#27579) needs to be fixed.

### PR DESCRIPTION
**Pull request by YHNdnzj**
Follow-up for https://github.com/systemd/systemd/pull/27579

In https://github.com/systemd/systemd/pull/27579 we refused all StartUnit requests for Type=dbus units
if dbus is not running, which means if dbus is manually stopped,
user can't use systemctl to start Type=dbus units again, which
is incorrect.

The only culprit that leads to the cancellation of the whole
transaction mentioned in https://github.com/systemd/systemd/issues/26799 is job type conflict on dbus.
So let's relax the restriction and only refuse job enqueuing
if dbus has a stop job.

To summarize, the case we want to avoid is:

1. dbus has a stop job installed
2. StartUnit/ActivationRequest is received
3. Type=dbus service gets started, which has Requires=dbus.socket
4. dbus is pulled in again, resulting in job type conflict

What we can support is:

1. dbus is already stopped
2. StartUnit is received (possibly through systemctl, i.e. on private bus)
3. Type=dbus service gets started
4. dbus is started again, thus the Type=dbus service

Fixes https://github.com/systemd/systemd/issues/27588

### CODE REVIEW

**Comment by YHNdnzj** [comment]
cc @Geass-LL

**Comment by Geass-LL** [comment]
The code is good to me. But the comment is not.

> On the other hand, if the stop job is already running, or dbus is already inactive, everything should be fine. 

unit_stop_pending only checks if there is a stop job, it doesn't care if this job is running or pending. 

> What we can support is:
> dbus is already stopping (in UNIT_DEACTIVATING)

This line should be "dbus is already stopped". Actually, when we stop a service manually,  at the time the service enters UNIT_DEACTIVATING state, `unit_stop_pending` returns true.

File: src/core/dbus-unit.c [hunk 1]
```diff
@@ -1877,11 +1877,25 @@ int bus_unit_queue_job(
...
-                return sd_bus_error_set(error, BUS_ERROR_SHUTTING_DOWN, "Refusing activation, D-Bus is not running.");
+        if (type == JOB_START && u->type == UNIT_SERVICE && SERVICE(u)->type == SERVICE_DBUS) {
+                /* dbus-broker issues StartUnit for activation requests, and Type=dbus services automatically
+                 * gain dependency on dbus.socket. Therefore, if dbus has a pending stop job, the new start
+                 * job that pulls in dbus again would result in job type conflict. */
+
+                Unit *dbus;
+                bool refuse = false;
+
+                dbus = manager_get_unit(u->manager, SPECIAL_DBUS_SOCKET);
```
Comment (dtardon) [inline src/core/dbus-unit.c]: You can use a loop here:
```c
FOREACH_STRING(unit, SPECIAL_DBUS_SOCKET, SPECIAL_DBUS_SERVICE) {
        dbus = manager_get_unit(u->manager, unit);
        if (dbus && unit_stop_pending(dbus))
                refuse = true;
}
```

File: src/core/dbus-unit.c [hunk 2]
```diff
@@ -1877,11 +1877,24 @@ int bus_unit_queue_job(
...
+                 * job that pulls in dbus again would result in job type conflict. */
+
+                bool refuse = false;
+
+                FOREACH_STRING(dbus_unit, SPECIAL_DBUS_SOCKET, SPECIAL_DBUS_SERVICE) {
+                        Unit *dbus;
+
+                        dbus = manager_get_unit(u->manager, dbus_unit);
+                        if (dbus && unit_stop_pending(dbus))
+                                refuse = true;
```
Comment (yuwata) [inline src/core/dbus-unit.c]:
```c
if (...) {
        refuse = true;
        break;
}
```

File: src/core/dbus-unit.c [hunk 3]
```diff
@@ -1877,11 +1877,24 @@ int bus_unit_queue_job(
...
 
-        /* dbus-broker issues StartUnit for activation requests, so let's apply the same check
-         * used in signal_activation_request(). */
-        if (type == JOB_START && u->type == UNIT_SERVICE &&
-            SERVICE(u)->type == SERVICE_DBUS && !manager_dbus_is_running(u->manager))
-                return sd_bus_error_set(error, BUS_ERROR_SHUTTING_DOWN, "Refusing activation, D-Bus is not running.");
+        if (type == JOB_START && u->type == UNIT_SERVICE && SERVICE(u)->type == SERVICE_DBUS) {
+                /* dbus-broker issues StartUnit for activation requests, and Type=dbus services automatically
+                 * gain dependency on dbus.socket. Therefore, if dbus has a pending stop job, the new start
+                 * job that pulls in dbus again would result in job type conflict. */
```
Comment (yuwata) [inline src/core/dbus-unit.c]: Please comment why `manager_dbus_is_running()` or `unit_inactive_or_pending()` cannot be used.

**Review by yuwata — COMMENTED** [review]
Why `signal_activation_request()` is not updated?
If the current behavior in `signal_activation_request()` can be kept, that is, `manager_dbus_is_running()` can be used, please add comment there.

File: src/core/dbus-unit.c (hunk 2 above)
Comment (dtardon) [reply src/core/dbus-unit.c]: Even better to return the dbus error directly from here.

Comment (yuwata) [reply src/core/dbus-unit.c]: Yeah, that's better.

**Review by yuwata — APPROVED** [review]

**Comment by YHNdnzj** [comment]
Force pushed to retrigger CI, plus some adjustments to the error message :)

**Review by bluca — APPROVED** [review]
